### PR TITLE
Automatically save generated passwords

### PR DESCRIPTION
**Pull request by samsymons**
Please review the release process for BrowserServicesKit [here](https://app.asana.com/0/1200194497630846/1200837094583426).

**Required**:

Task/Issue URL: https://app.asana.com/0/0/1201969773101195/f
iOS PR:  https://github.com/duckduckgo/iOS/pull/1109
macOS PR: https://github.com/more-duckduckgo-org/macos-browser/pull/500

**Description**:

This PR updates BSK to automatically save generated credentials when the user selects that option. Note that this PR does not require any changes on either of the client codebases.

If the user doesn't have a username entered, they will get prompted immediately. If they _have_ already entered a username, the credential gets saved without prompting them. See [this task](https://app.asana.com/0/1175293949586521/1202076731691825/f) for context.

**Steps to test this PR**:
1. Visit a site that has a sign-up form; I've been using [Hacker News](http://news.ycombinator.com/login) since it's a super simple page
1. Click into the password field, and click the generated password prompt, and ensure you have no username
1. You should get a prompt after clicking the generated password field; go ahead and save that
1. Go back to the sign up form, type in a username, and hit enter (it's fine to use an existing username to prevent creating an actual account, the app just needs the form submission event to trigger)
1. Check the Autofill panel and verify that the previous login has been updated to include the username

**OS Testing**:

* [ ] iOS 13
* [ ] iOS 14
* [ ] iOS 15
* [ ] macOS 10.15
* [ ] macOS 11

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

### CODE REVIEW

**Comment by samsymons** [comment]
Marking `DO NOT MERGE` temporarily, since this PR additionally needs approval from the frontend team.

File: Sources/BrowserServicesKit/SecureVault/SecureVaultManager.swift [hunk 1]
```diff
@@ -142,17 +152,94 @@ extension SecureVaultManager: AutofillSecureVaultDelegate {
...
+        // As a precaution, check whether an account exists with the matching generated password _and_ a non-nil username.
+        // If so, then the user must have already saved the generated credentials and set a username.
+        
+        for account in accounts where !account.username.isEmpty {
+            if let accountID = account.id,
+               let credentialsForAccount = try vault.websiteCredentialsFor(accountId: accountID),
+               credentialsForAccount.password == passwordData,
+               account.username == autogeneratedCredentials.username ?? "" {
+                os_log("Tried to save autogenerated password but it already exists; not saving credentials", log: .passwordManager)
+                return
```
Comment (THISISDINOSAUR) [inline Sources/BrowserServicesKit/SecureVault/SecureVaultManager.swift:188]: Nit: potentially controversial indentation

**Review by THISISDINOSAUR — APPROVED** [review]
LGTM! 
(fyi we chatted about some parts of this on MM)

File: Sources/BrowserServicesKit/Autofill/AutofillUserScript+SecureVault.swift [hunk 2]
```diff
@@ -134,30 +134,36 @@ extension AutofillUserScript {
     
     // MARK: - Requests
     
-    public struct IncomingCredentials {
+    public struct IncomingCredentials: Equatable {
         
+        // swiftlint:disable:next nesting
```
[the remark this answers is not in the document]
Comment (samsymons) [reply Sources/BrowserServicesKit/Autofill/AutofillUserScript+SecureVault.swift:139]: That's... a very good question, I would say no. I'll consult the style guide, and likely disable this at the configuration level.

File: Sources/BrowserServicesKit/SecureVault/SecureVaultManager.swift (hunk 1 above)
Comment (samsymons) [reply Sources/BrowserServicesKit/SecureVault/SecureVaultManager.swift:188]: Oh yeah this is no good, let me fix

**Comment by GioSensation** [comment]
Updated autofill to tag 4.3.0.

**Pull request merged**